Source/Scene/ModelExperimental/PickingPipelineStage.js

```diff
@@ -0,0 +1,123 @@
+import Color from "../../Core/Color.js";
+import ComponentDatatype from "../../Core/ComponentDatatype.js";
+import defined from "../../Core/defined.js";
+import Buffer from "../../Renderer/Buffer.js";
+import BufferUsage from "../../Renderer/BufferUsage.js";
+import ShaderDestination from "../../Renderer/ShaderDestination.js";
+
+/**
+ * The picking pipeline stage is responsible for handling picking of primitives.
+ *
+ * @namespace PickingPipelineStage
+ * @private
+ */
+var PickingPipelineStage = {};
+
+/**
+ * Process a primitive. This modifies the following parts of the render resources:
+ * <ul>
+ *  <li>adds attribute and varying declaration for the pick color vertex attribute in the vertex shader for instanced meshes</li>
+ *  <li>adds declaration for the pick color uniform for non-instanced meshes</li>
+ *  <li>adds defines in the shader for when picking is enabled</li>
+ *  <li>creates the pick ID objects in the context</li>
+ * </ul>
+ * @param {PrimitiveRenderResources} renderResources The render resources for this primitive.
+ * @param {ModelComponents.Primitive} node The primitive.
+ * @param {FrameState} frameState The frame state.
+ */
+PickingPipelineStage.process = function (
+  renderResources,
+  primitive,
+  frameState
+) {
+  var context = frameState.context;
+  var runtimeNode = renderResources.runtimeNode;
+  var shaderBuilder = renderResources.shaderBuilder;
+
+  shaderBuilder.addDefine("ALLOWS_PICKING");
```

Comment:
To be consistent with other pipeline stages:

```suggestion
  shaderBuilder.addDefine("USE_PICKING");
```